httputil/transport.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2015 The Go Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file or at
+// https://developers.google.com/open-source/licenses/bsd.
+
+// This file implements a http.RoundTripper that authenticates
+// requests issued against api.github.com endpoint.
+
+package httputil
+
+import (
+	"log"
+	"net/http"
+	"os"
+
+	"google.golang.org/cloud/compute/metadata"
+)
+
+type Transport struct {
+	UserAgent    string
+	Token        string
+	ClientID     string
+	ClientSecret string
+	Base         http.RoundTripper
+}
+
+func NewTransport(base http.RoundTripper) *Transport {
+	if metadata.OnGCE() {
+		return NewTransportFromMetadata(base)
+	}
+	return NewTransportFromEnvironment(base)
+}
+
+func NewTransportFromEnvironment(base http.RoundTripper) *Transport {
+	return &Transport{
+		UserAgent:    os.Getenv("USER_AGENT"),
+		Token:        os.Getenv("GITHUB_TOKEN"),
+		ClientID:     os.Getenv("GITHUB_CLIENT_ID"),
+		ClientSecret: os.Getenv("GITHUB_CLIENT_SECRET"),
+		Base:         base,
+	}
+}
+
+func NewTransportFromMetadata(base http.RoundTripper) *Transport {
+	return &Transport{
+		UserAgent:    gceAttr("user-agent"),
+		Token:        gceAttr("github-token"),
+		ClientID:     gceAttr("github-client-id"),
+		ClientSecret: gceAttr("github-client-secret"),
+		Base:         base,
+	}
+}
+
+func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if t.UserAgent != "" {
```

Comment:
RoundTrippers aren't supposed to mutate the provided request.
What you need to do here is something like `cloneRequest` in this file https://github.com/golang/oauth2/blob/master/transport.go#L96 before you mutate the headers.